core/src/main/java/com/linecorp/armeria/common/http/AggregatedHttpMessage.java

```diff
@@ -205,4 +229,80 @@ default String authority() {
     default HttpStatus status() {
         return headers().status();
     }
+
+    /**
+     * Converts this message into a new complete {@link HttpRequest}.
+     *
+     * @return the converted {@link HttpRequest} whose stream is complete
+     * @throws IllegalStateException if this message is not a request
+     */
+    default HttpRequest toHttpRequest() {
+        final HttpHeaders headers = headers();
+
+        // From the section 8.1.2.3 of RFC 7540:
+        //// All HTTP/2 requests MUST include exactly one valid value for the :method, :scheme, and :path
+        //// pseudo-header fields, unless it is a CONNECT request (Section 8.3)
+        // NB: ':scheme' will be filled when a request is written.
+        if (headers.method() == null) {
+            throw new IllegalStateException("not a request (missing :method)");
+        }
+        if (headers.path() == null) {
+            throw new IllegalStateException("not a request (missing :path)");
+        }
+
+        final DefaultHttpRequest req = new DefaultHttpRequest(headers);
+        final HttpData content = content();
+        if (!content.isEmpty()) {
+            req.headers().setInt(HttpHeaderNames.CONTENT_LENGTH, content.length());
+            req.write(content);
+        }
+
+        final HttpHeaders trailingHeaders = trailingHeaders();
+        if (!trailingHeaders.isEmpty()) {
+            req.write(trailingHeaders);
+        }
+
+        req.close();
+        return req;
+    }
+
+    /**
+     * Converts this message into a new complete {@link HttpResponse}.
+     *
+     * @return the converted {@link HttpResponse} whose stream is complete
+     * @throws IllegalStateException if this message is not a response.
+     */
+    default HttpResponse toHttpResponse() {
+        final DefaultHttpResponse res = new DefaultHttpResponse();
+        final List<HttpHeaders> informationals = informationals();
+        final HttpHeaders headers = headers();
+
+        // From the section 8.1.2.4 of RFC 7540:
+        //// For HTTP/2 responses, a single :status pseudo-header field is defined that carries the HTTP status
+        //// code field (see [RFC7231], Section 6). This pseudo-header field MUST be included in all responses;
+        //// otherwise, the response is malformed (Section 8.1.2.6).
+        if (headers.status() == null) {
+            throw new IllegalStateException("not a response (missing :status)");
+        }
+
+        if (!informationals.isEmpty()) {
+            informationals.forEach(res::write);
+        }
+
+        final HttpData content = content();
+        if (!content.isEmpty()) {
+            headers.setInt(HttpHeaderNames.CONTENT_LENGTH, content.length());
```

Comment:
It's HTTP/2, so I guess `transfer-encoding` header is of no use. Internally, Armeria will remove it before sending it via HTTP/2 and will add it if necessary when sending via HTTP/1.